### ISSUE TITLE
[MODULES-63] Allow module activation trigger on first class load

### ISSUE THREAD

**Issue by dmlloyd** [jboss-modules/jboss-modules#479]
@tdiesler opened **[MODULES-63](https://issues.redhat.com/browse/MODULES-63?redirect=false)** and commented

See 4.2. Core Specification
4.4.6.2 Lazy Activation Policy

The lazy activation policy allows a Framework implementation to defer the
creation of the bundle class loader and activation of the bundle until the
bundle is first used; potentially saving resources and initialization time dur-
ing startup.

By default, any class load can trigger the lazy activation, however, resource
loads must not trigger the activation. The lazy activation policy can define
which classes cause the activation with the following directives:

* include - A list of package names that must trigger the activation when
  a class is loaded from any of these packages. The default is all package
  names present in the bundle.

* exclude - A list of package names that must not trigger the activation of
  the bundle when a class is loaded from any of these packages. The default
  is no package names.

When a class load triggers the lazy activation, the Framework must first
define the triggering class. This definition can trigger additional lazy activa-
tions. These activations must be deferred until all transitive class loads and
defines have finished. Thereafter, the activations must be executed in the
reverse order of detection. That is, the last detected activation must be exe-
cuted first. Only after all deferred activations are finished must the class
load that triggered the activation return with the loaded class. If an error
occurs during this process, it should be reported as a Framework ERROR
event. However, the class load must succeed normally. A bundle that fails its
lazy activation should not be activated again until the framework is
restarted or the bundle is explicitly started by calling the Bundle start
method.

---
No further details from [MODULES-63](https://issues.redhat.com/browse/MODULES-63?redirect=false)

**Comment by dmlloyd** [comment]
@tdiesler commented

This can possibly be done by simply hooking into the module's local class load

https://github.com/jbosgi/jboss-modules/commit/ce29f57fd0432f8bf69076f89321d72d8ecb6b5e

**Comment by dmlloyd** [comment]
@dmlloyd commented

The intended way to accomplish this is to only put the "eager" classes in your primary resource loader.  Then use a fallback loader or auxiliary local loader dependency for the "lazy" packages.  When the lazy package is accessed, modify your primary resource loader to add the new package, recalculate resource paths, and relink the module, then recursively re-request the class and it should be there.

**Comment by dmlloyd** [comment]
@tdiesler commented

Done

This is done in the OSGi layer

**Comment by dmlloyd** [comment]
@dmlloyd commented

Removing from release notes.
